test/plugins/kubernetes.js

```diff
@@ -0,0 +1,109 @@
+const MockController = require("../mocks/controller");
+const K8sHelper = require("../mocks/kubernetes");
+const { Client } = require("kubernetes-client");
+
+const kubernetes = require("../../plugins/kubernetes");
+
+describe("plugins / kubernetes", () => {
+    let controller;
+
+    describe("Setup", () => {
+        describe("getKubernetesClient", () => {
+            it("should return without error", () => {
+                kubernetes.getK8s();
+            });
+        });
+    });
+
+    describe("Methods", () => {
+        beforeEach(() => {
+            controller = new MockController({});
+            this.stream = new Readable({ read() {} });
+            this.client = new Client({ backend: {}, version: "1.13" });
+            K8sHelper.addGroup(this.client, "events.k8s.io");
+
+            this.callback = sinon.spy();
+            const jsonStream = new JSONStream();
+            this.stream.pipe(jsonStream);
+            this.getObjectStream = sinon.stub(
+                this.client.api.v1.watch.namespaces().event,
+                "getObjectStream"
+            );
+            this.getObjectStream.resolves(jsonStream);
+            kubernetes.getK8s = sinon.stub().returns(this.client);
+        });
+
+        describe("init", () => {
+            it("should add plugin extensions", () => {
+                kubernetes.init(controller);
+                sinon.assert.calledOnce(controller.addPluginExtension);
+                sinon.assert.calledWith(
+                    controller.addPluginExtension,
+                    "kubernetes",
+                    kubernetes
+                );
+            });
+        });
+
+        it("watching: handles specified resources", async () => {
+            const cb = sinon.spy();
+            const kubeResource = {
+                type: "event",
+                resource: "",
+                group: "",
+            };
+            const testObject =
+                '{"type": "ADDED", "object": {"metadata": {"creationTimestamp": "3000-07-16T18:44:46Z"}}}';
+
+            const stop = kubernetes.startWatch(
+                kubeResource,
+                "test-namespace",
+                cb
+            );
+            this.stream.push(testObject);
+            await stop();
+            sinon.assert.called(cb);
+        });
+        it("watching: handle specified resources", async () => {
+            const cb = sinon.spy();
+            const kubeResource = {
+                type: "event",
+                version: "v1",
+                group: "events.k8s.io",
+            };
+            const testObject =
+                '{"type": "ADDED", "object": {"metadata": {"creationTimestamp": "3000-07-16T18:44:46Z"}}}';
+
+            const stop = kubernetes.startWatch(
+                kubeResource,
+                "test-namespace",
+                cb
+            );
+            this.stream.push(testObject);
+            await stop();
+            sinon.assert.called(cb);
```

Comment:
can we also have some assertions that expect the `cb` was not called when working with resources we aren't currently watching?

Reply:
Since we are receiving a stream of events for a specific resource directly from the kubernetes-client, we can be sure that we won't receive conflicting resource event types.